Volodymyr Lut, UCU 2019

### Data overview

In this project the data from the [Ukrainian Open Data Portal](https://data.gov.ua) was used.

The dataset contains information about international bus routes and bus stops in Ukraine. Under [this link](https://data.gov.ua/dataset/d5d79cf2-ccc7-40a9-9b7f-4c5cd0d061f9) two files are allocated: `international_routes_stops3.csv` (contains information about stops and schedule of international routes) and `international_routes.csv` (contains information about starting and destination points of the routes, along with schedules). Copies of those datasets (actual on Monday, June 10 2019) are contained under the `raw_data` directory of this repository.

In [1]:
import pandas as pd

In [2]:
stops = pd.read_csv('./raw_data/international_routes_stops3.csv', sep=';', encoding='windows-1251')
stops.head()

id rejs_number stop_city_name                             stop_address  \
0  CZ1357           1           Київ  вулиця Набережно-Печерська дорога, 10-а   
1  CZ1357           1           Київ                       вул. С.Петлюри, 32   
2  CZ1357           1        Житомир                        вул. Київська, 93   
3  CZ1357           1          Рівне                        вул. Київська, 40   
4  CZ1357           1          Львів                      вул. Стрийська, 109   

    stop_code route_frequency distance_direct arrival_direct departure_direct  \
0  3200000000         Щоденно               0          20:00            20:10   
1  3200000000         Щоденно              12          20:30            21:00   
2  1810100000         Щоденно             140          23:20            23:40   
3  5610100000         Щоденно             328           2:50             3:10   
4  4610100000         Щоденно             539           6:10             6:30   

  time_diff_direct arrival_return departure_return time_diff_return  \
0                0          15:30              NaN                0   
1                0          14:40            15:00                0   
2                0          12:50            13:00                0   
3                0          10:30            10:40                0   
4                0           8:35             8:50                0   

   stop_number  Unnamed: 14  
0          0.0          NaN  
1          1.0          NaN  
2          2.0          NaN  
3          3.0          NaN  
4          4.0          NaN

In [3]:
routes = pd.read_csv('./raw_data/international_routes.csv')
routes.head()

permision_delivery_date  n/o    id  \
0                     NaN    3   MD3   
1                     NaN    8   MD8   
2              05.06.2018   11  RU11   
3                     NaN   12  RU12   
4                     NaN   13  RU13   

                                    march_route_name  \
0  АК АЕРОПОРТ БОРИСПІЛЬ (вул. Бориспільська, 7) ...   
1            Балта АС (вул. Уварова, 38а) - Дубосари   
2                                      Бахмут-Москва   
3  Бахмут (АС вул. Космонавтів,5)-Москва (Новоясе...   
4                                    Бахму - Москва    

                                        company_name  \
0                                     ФОП Гергі Є.М.   
1  ПП Фаворіт 65005, м. Одеса, вул. Бугаївська, 2...   
2                 ТОВ"Форум Авто"|ФОП"Мелкумян Ю.О."   
3                                   ТОВ "СЕВЕР-АВТО"   
4                              ФОП Базілєвський В.О.   

                               company_name_external border_crossing  \
0                            S.R.L. "SARVALTEH AUTO"          Табаки   
1  ОАО "Дубоссарское автотранспортное предприятие...       Платоново   
2                                 ООО"Газавторемонт"        Гоптівка   
3                                       без паритету      Плетенівка   
4                                  ООО "Горизон-Тур"       Танюшівка   

  border_crossing_alternative march_route_time_direct march_route_time_return  \
0                       16:00                   16:00                     NaN   
1                         NaN                     NaN                     NaN   
2                  Плетенівка                   16:50                   16:50   
3                         NaN                     NaN                     NaN   
4                         NaN                   19:30                   15:30   

  march_route_regularity        license_terms destionation_country_name  \
0                Щоденно  2020-02-08 00:00:00                   Молдова   
1                    NaN  2020-06-07 00:00:00                   Молдова   
2                Щоденно  2019-07-20 00:00:00                     Росія   
3                    NaN  2019-02-28 00:00:00                     Росія   
4                Щоденно  2019-01-07 00:00:00                     Росія   

  destionation_country_code  march_route_status  
0                        MD                True  
1                        MD                True  
2                        RU                True  
3                        RU                True  
4                        RU                True

In this project file `international_routes_stops3.csv` will be used. It should be filtered - in this research I'm interested only in routes, which are heading to Poland. 

**Explanation of useful columns:**

1. *ID* decodes the bus line id - when checking this ID in the `international_routes.csv` dataset, one could determine line operator's comoany name, information about it's license, border crossing point etc.

2. *stop_city_name* together with *stop_address* will be used to determine geolocation of the stop, which will be used for visualisation on maps.

3. *stop_number* is a column which will be used to determine the position of the stop in the route.

In [4]:
### Filtering out Polish routes
pol = stops[stops.id.str[:2].str.upper() == 'PL']
pol = pol.dropna(subset=['stop_city_name', 'stop_address'])
pol.shape

(6446, 15)

In [5]:
import numpy as np

In [6]:
np.unique(pol.id).shape

(482,)

There are **6446 stops** belonging to **482 bus lines** heading to Poland after clearing the data.

### Data Mining

For visualizations we need to know the exact location of every bus stop. In order to scrape this data, [Google Maps API](https://developers.google.com/maps) will be used.

In [7]:
pol = pol.drop("Unnamed: 14", axis=1)
pol = pol.assign(full_address=(pol.stop_city_name + ', ' + pol.stop_address).values)
pol.head()

id rejs_number       stop_city_name      stop_address   stop_code  \
2530  PL944           1               Жовква  пл. Коновальця 1  4622710100   
2531  PL944           1          Рава Руська  вул. Двірцева, 4  4622710400   
2534  PL944           1  Томашів Любельський   ul. Zamojska, 9      22-600   
2535  PL944           2               Жовква  пл. Коновальця 1  4622710100   
2536  PL944           2          Рава Руська  вул. Двірцева, 4  4622710400   

     route_frequency distance_direct arrival_direct departure_direct  \
2530         Щоденно               0           4:20             4:40   
2531         Щоденно              32           5:25             5:30   
2534         Щоденно              63           6:05              NaN   
2535         Щоденно               0          21:50            22:10   
2536         Щоденно              32          22:55            23:00   

     time_diff_direct arrival_return departure_return time_diff_return  \
2530                0          14:55              NaN                0   
2531                0          14:05            14:10                0   
2534               -1          11:00            11:30                0   
2535                0           6:25              NaN                0   
2536                0           5:35             5:40                0   

      stop_number                          full_address  
2530          0.0              Жовква, пл. Коновальця 1  
2531          1.0         Рава Руська, вул. Двірцева, 4  
2534          4.0  Томашів Любельський, ul. Zamojska, 9  
2535          5.0              Жовква, пл. Коновальця 1  
2536          6.0         Рава Руська, вул. Двірцева, 4

In [8]:
np.unique(pol.full_address).shape

(2080,)

In [37]:
import googlemaps
import re
# Uncomment line below and paste your key there
# gmaps = googlemaps.Client(key='<YOUR KEY HERE>')

In [9]:
from tqdm import tqdm

In [73]:
# Use this code to scrape geocoding data and save it to file
geocoded_addresses = []
for address in tqdm(np.unique(pol.full_address)):
    geocode_result = gmaps.geocode(address)
    geocoded_addresses.append(geocode_result)
np.savetxt("geocoded.tmp", geocoded_addresses, fmt='%s')

100%|██████████| 2080/2080 [18:28<00:00,  1.77it/s]


In [67]:
import json

In [131]:
# Use this code to restore geocoding data from file
with open('geocoded.tmp') as fp:  
    geocoded_addresses = []   
    geocode_result = fp.readline()
    i = 0
    while geocode_result:
        geocode_result = geocode_result.replace("Boryspil'", "Boryspil")
        geocode_result = re.sub("(?<=[a-z])'(?=[a-z])", "", geocode_result, flags=re.IGNORECASE).replace("True", '"true"').replace("False", '"false"').replace("'", '"').replace('""', '"').replace('"-', '-')
        geocode_result = re.sub('(?<=[a-z]) "(?=[a-z])', " ", geocode_result, flags=re.IGNORECASE)
        geocode_result = re.sub('(?<=[a-z])" (?=[a-z])', " ", geocode_result, flags=re.IGNORECASE)
        geocode_result = re.sub('(?<=[a-z])", (?=[a-z])', " ", geocode_result, flags=re.IGNORECASE)
        if(i == 735):
            print(geocode_result)
        geocoded_addresses.append(json.loads(geocode_result))
        geocode_result = fp.readline()

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

1840
1841
1842
1843
1844
1845
1846
1847
1848
1849
1850
1851
1852
1853
1854
1855
1856
1857
1858
1859
1860
1861
1862
1863
1864
1865
1866
1867
1868
1869
1870
1871
1872
1873
1874
1875
1876
1877
1878
1879
1880
1881
1882
1883
1884
1885
1886
1887
1888
1889
1890
1891
1892
1893
1894
1895
1896
1897
1898
1899
1900
1901
1902
1903
1904
1905
1906
1907
1908
1909
1910
1911
1912
1913
1914
1915
1916
1917
1918
1919
1920
1921
1922
1923
1924
1925
1926
1927
1928
1929
1930
1931
1932
1933
1934
1935
1936
1937
1938
1939
1940
1941
1942
1943
1944
1945
1946
1947
1948
1949
1950
1951
1952
1953
1954
1955
1956
1957
1958
1959
1960
1961
1962
1963
1964
1965
1966
1967
1968
1969
1970
1971
1972
1973
1974
1975
1976
1977
1978
1979
1980
1981
1982
1983
1984
1985
1986
1987
1988
1989
1990
1991
1992
1993
1994
1995
1996
1997
1998
1999
2000
2001
2002
2003
2004
2005
2006
2007
2008
2009
2010
2011
2012
2013
2014
2015
2016
2017
2018
2019
2020
2021
2022
2023
2024
2025
2026
2027
2028
2029
2030
2031
2032
2033
2034
2035
2036
2037
2038
2039


In [125]:
mem_geocoded_addresses = geocoded_addresses

In [132]:
geocoded_addresses

[[{'address_components': [{'long_name': '30',
     'short_name': '30',
     'types': ['street_number']},
    {'long_name': 'Zaliznychna Street',
     'short_name': 'Zaliznychna St',
     'types': ['route']},
    {'long_name': 'Ivano-Frankivsk',
     'short_name': 'Ivano-Frankivsk',
     'types': ['locality', 'political']},
    {'long_name': 'Ivano-Frankivska city council',
     'short_name': 'Ivano-Frankivska city council',
     'types': ['administrative_area_level_3', 'political']},
    {'long_name': 'Ivano-Frankivska oblast',
     'short_name': 'Ivano-Frankivska oblast',
     'types': ['administrative_area_level_1', 'political']},
    {'long_name': 'Ukraine',
     'short_name': 'UA',
     'types': ['country', 'political']},
    {'long_name': '76000', 'short_name': '76000', 'types': ['postal_code']}],
   'formatted_address': 'Zaliznychna St, 30, Ivano-Frankivsk, Ivano-Frankivska oblast, Ukraine, 76000',
   'geometry': {'location': {'lat': 48.9220657, 'lng': 24.7247958},
    'location_

In [123]:
re.sub('(?<=[a-z]) "(?=[a-z])', "", 'Avtostantsiya "Vydubychi', flags=re.IGNORECASE)

'AvtostantsiyaVydubychi'